### PR TITLE
EVP: Only use the engine when one is defined, in pkey_mac_ctrl()

### LINKED ISSUE #11671: Null pointer dereference in 3.0.0-alpha1 when generating a CMAC key
What I think is the correct way of creating a CMAC EVP_PKEY (from the wiki: https://wiki.openssl.org/index.php/EVP_Key_and_Parameter_Generation) now hits a segfault trying to dereference the null ENGINE pointer in the EVP_PKEY_CTX:

```c
include <openssl/evp.h>

int main() {
	EVP_PKEY_CTX *kctx = EVP_PKEY_CTX_new_id(EVP_PKEY_CMAC, NULL);
	EVP_PKEY_keygen_init(kctx);
	EVP_PKEY_CTX_ctrl(kctx, -1, EVP_PKEY_OP_KEYGEN, EVP_PKEY_CTRL_CIPHER, 0, (void *)EVP_aes_256_ecb());
}
```

```
rogram terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000000000516c5d in ENGINE_get_id (e=0x0) at crypto/engine/eng_lib.c:251
251	    return e->id;
Missing separate debuginfos, use: dnf debuginfo-install glibc-2.30-11.fc31.x86_64
(gdb) bt
#0  0x0000000000516c5d in ENGINE_get_id (e=0x0) at crypto/engine/eng_lib.c:251
#1  0x000000000053d34e in pkey_mac_ctrl (ctx=0x12d1080, type=12, p1=0, p2=0x70a7a0 <aesni_256_ecb>) at crypto/evp/pkey_mac.c:312
#2  0x0000000000413ce6 in EVP_PKEY_CTX_ctrl (ctx=0x12d1080, keytype=-1, optype=4, cmd=12, p1=0, p2=0x70a7a0 <aesni_256_ecb>)
    at crypto/evp/pmeth_lib.c:978
#3  0x0000000000404139 in main ()
```

### PR DESCRIPTION
**Pull request by levitte**
Fixes #11671

### CODE REVIEW

**Comment by levitte** [comment]
> Oops

Yup

**Comment by mattcaswell** [comment]
Can we have a test (perhaps in evp_extra_test) for creating a CMAC key as per #11671?

**Comment by levitte** [comment]
I finally got around to add a quick test

**Comment by levitte** [comment]
Rebased to get rid of the mdl failure.

File: test/evp_extra_test.c [hunk 1]
```diff
@@ -1179,6 +1179,27 @@ static int test_EVP_PKEY_check(int i)
...
+static int test_CMAC_keygen(void)
+{
+    /*
+     * This is a legacy method for CMACs, but should still work.
+     * This verifies that it works without an ENGINE.
+     */
+    EVP_PKEY_CTX *kctx = EVP_PKEY_CTX_new_id(EVP_PKEY_CMAC, NULL);
+    int ret = 0;
+
+    if (!TEST_true(EVP_PKEY_keygen_init(kctx) > 0)
```
Comment (slontis) [inline test/evp_extra_test.c:1191]: This can return negative if the kctx is NULL..

**Comment by levitte** [comment]
Merged

bcb018e70b596811146a17d95532498bbe3ef13a EVP: Only use the engine when one is defined, in pkey_mac_ctrl()
a87820e16bbbbb1f8a68ddaf3aa1159da886acca test/evp_extra_test.c: Add test for CMAC keygen with a NULL engine

**Comment by mattcaswell** [comment]
This seems to have broken Travis.

**Comment by levitte** [comment]
How the hell that happened...
Will look.

**Comment by levitte** [comment]
Fix in #11833

**Comment by anamulhkp** [comment]
ok

বৃহস্পতি, ১৪ মে, ২০২০ ৯:৩৩ PM তারিখে Richard Levitte <
notifications@github.com> লিখেছেন:

**Pull request closed without merging**